src/rust/engine/process_execution/src/sandboxer.rs

```diff
@@ -0,0 +1,290 @@
+        let socket_path = self.socket_path.clone();
+        let idle_beats = Arc::new(AtomicUsize::new(0));
+
+        let idle_beats_ref = idle_beats.clone();
+        thread::spawn(move || loop {
+            thread::sleep(SandboxerService::POLLING_INTERVAL);
+            if !fs::exists(&socket_path).unwrap_or(false) {
+                warn!("Socket file {} deleted. Exiting.", &socket_path.display());
+                process::exit(22);
+            }
+            if idle_beats_ref.fetch_add(1, Ordering::Relaxed) > Self::ALLOWED_IDLE_BEATS {
+                warn!("Idle time limit exceeded. Exiting.");
+                process::exit(23);
+            }
+        });
+
+        info!(
+            "Starting sandboxer server on socket path {}",
+            self.socket_path.display()
+        );
+        Server::builder()
+            .add_service(SandboxerGrpcServer::new(SandboxerGrpcImpl {
+                inactivity_counter: idle_beats,
+            }))
+            .serve_with_incoming(UnixListenerStream::new(listener))
+            .await?;
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub struct SandboxerGrpcImpl {
+    // We increment this at every poll interval, and reset it on every gRPC call.
+    inactivity_counter: Arc<AtomicUsize>,
+}
+
+#[tonic::async_trait]
+impl SandboxerGrpc for SandboxerGrpcImpl {
+    async fn copy_local_file(
+        &self,
+        request: Request<CopyLocalFileRequest>,
+    ) -> Result<Response<CopyLocalFileResponse>, Status> {
+        self.inactivity_counter.store(0, Ordering::Relaxed);
+        let r = request.into_inner();
+        // NB: It's important that we actually copy, rather than move, the file, as
+        // we must create a new inode. We copy first to a tmp destination and then
+        // rename atomically, so that the final path is never in a half-copied state.
+        // We reasonably assume that the tmp destination is not occupied: The suffix
+        // is obscure, and only one sandboxer should ever be writing into a given sandbox.
+        let tmp_dst = format!("{}._sandboxer.dst.tmp", &r.dst);
+        let nbytes = fs::copy(&r.src, &tmp_dst).map_err(|e| {
+            let err_msg = e.to_string();
+            warn!("Error copying {} to {}: {}", &r.src, &tmp_dst, err_msg);
+            Status::failed_precondition(err_msg)
+        })?;
+        fs::rename(&tmp_dst, &r.dst).map_err(|e| {
+            let err_msg = e.to_string();
+            warn!("Error remaning {} to {}: {}", &tmp_dst, &r.dst, err_msg);
+            Status::failed_precondition(err_msg)
+        })?;
+
+        let confirmation = format!("Copied {} bytes from {} to {}", nbytes, &r.src, &r.dst);
+        info!("{}", confirmation);
+        Ok(Response::new(CopyLocalFileResponse {
+            confirmation: { confirmation },
+        }))
+    }
+}
+
+// The client that sends requests to the sandboxer service.
+pub struct SandboxerClient {
```

Comment:
Would it be useful to have a (minimal) "sandbox setup" trait so we can abstract over the sandbox setup strategy?